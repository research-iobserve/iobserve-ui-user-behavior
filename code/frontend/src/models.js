// @flow

import { List, Map } from "immutable";
import Record from "./types";
import type {
  ApplicationT,
  ApplicationStateT,
  PageT,
  PageStateT,
  VisitT,
  VisitStateT,
  SettingsStateT,
  LayoutT,
  StateT
} from "./types";

/**
 * Application
 */
const applicationDefaults: ApplicationT = {
  id: 0,
  name: "",
  description: "",
  custom: false
};
export class Application extends Record(applicationDefaults) {
  id: number;
  name: string;
  description: string;
  custom: boolean;
}

/**
 * Page
 */
const pageDefaults: PageT = {
  id: 0,
  name: "",
  extra: Map()
};
export class Page extends Record(pageDefaults) {
  id: number;
  name: string;
  extra: Map<string, string>;
}

/**
 * Visit
 */
const visitDefaults: VisitT = {
  id: 0,
  count: 0,
  start: new Page(),
  end: new Page(),
  action: ""
};
export class Visit extends Record(visitDefaults) {
  id: number;
  count: number;
  start: Page;
  end: Page;
  action: string;
}

/**
 * ApplicationState
 */
const applicationStateDefaults: ApplicationStateT = {
  applications: List()
};
export class ApplicationState extends Record(applicationStateDefaults) {
  applications: List<Application>;
}

/**
 * PageState
 */
const pageStateDefaults: PageStateT = {
  pages: Map()
};
export class PageState extends Record(pageStateDefaults) {
  pages: Map<number, List<Page>>;
}

/**
 * VisitState
 */
const visitStateDefaults: VisitStateT = {
  visits: Map()
};
export class VisitState extends Record(visitStateDefaults) {
  visits: Map<number, List<Visit>>;
}

/**
 * SettingsState
 */
const settingsStateDefaults: SettingsStateT = {
  layout: "breadthfirst"
};
export class SettingsState extends Record(settingsStateDefaults) {
  layout: LayoutT;
}

/**
 * State
 */
const stateDefaults: StateT = {
  application: new ApplicationState(),
  page: new PageState(),
  visit: new VisitState(),
  settings: new SettingsState()
};
export class State extends Record(stateDefaults) {
  application: ApplicationState;
  page: PageState;
  visit: VisitState;
  settings: SettingsState;
}
