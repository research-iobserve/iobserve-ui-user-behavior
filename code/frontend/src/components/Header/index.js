// @flow

import React from "react";
import { List } from "immutable";
import { Link } from "react-router";
import {
  Collapse,
  Navbar,
  NavbarToggler,
  NavbarBrand,
  Nav,
  NavItem,
  NavLink,
  NavDropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem
} from "reactstrap";

import type { LayoutT } from "../../types";
import { SettingsState } from "../../models";

type Props = {
  changeSettings: (string, LayoutT) => void,
  settings: SettingsState
};
type State = {
  isOpen: boolean,
  editorDropdown: boolean,
  settingsDropdown: boolean
};

class Header extends React.Component {
  props: Props;
  state: State;

  // TODO: combine with type
  layouts = List(["cose", "grid", "breadthfirst"]);

  toggle: (string) => () => void;

  constructor(props: Props) {
    super(props);

    this.toggle = what => () => this.setState({
      [what]: !this.state[what]
    });

    this.state = {
      isOpen: false,
      editorDropdown: false,
      settingsDropdown: false
    };
  }

  render() {
    return (
      <Navbar color="faded" light toggleable>
        <NavbarToggler right onClick={this.toggle("isOpen")} />
        <NavbarBrand tag={Link} to="/">iObserve UBM Visualization</NavbarBrand>
        <Collapse isOpen={this.state.isOpen} navbar>
          <Nav className="ml-auto" navbar>
            <NavItem>
              <NavLink tag={Link} to="/">Visualization</NavLink>
            </NavItem>
            <NavDropdown
              isOpen={this.state.editorDropdown}
              toggle={this.toggle("editorDropdown")}>
              <DropdownToggle nav caret>Editor</DropdownToggle>
              <DropdownMenu right>
                <DropdownItem>
                  <Link to="/editor/new">New Application</Link>
                </DropdownItem>
                <DropdownItem>
                  <Link to="/editor">All Applications</Link>
                </DropdownItem>
              </DropdownMenu>
            </NavDropdown>
            <NavDropdown
              isOpen={this.state.settingsDropdown}
              toggle={this.toggle("settingsDropdown")}>
              <DropdownToggle nav caret>Settings</DropdownToggle>
              <DropdownMenu right>
                <DropdownItem header>Layout</DropdownItem>
                {this.layouts.map((layout, key) => (
                  <DropdownItem
                    onClick={() => this.props.changeSettings("layout", layout)}
                    key={key}>
                    {layout === this.props.settings.get("layout")
                      ? <strong>{layout}</strong>
                      : layout}
                  </DropdownItem>
                ))}
              </DropdownMenu>
            </NavDropdown>
          </Nav>
        </Collapse>
      </Navbar>
    );
  }
}

export default Header;
