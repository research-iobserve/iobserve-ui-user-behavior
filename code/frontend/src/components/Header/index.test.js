// @flow

import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";

import Header from "./index";
import { SettingsState } from "../../models";

describe("<Header />", () => {
  it("should render", () => {
    const tree = renderer
      .create(
        <Header changeSettings={() => {}} settings={new SettingsState()} />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should open and close on toggle", () => {
    const renderedComponent = shallow(
      <Header changeSettings={() => {}} settings={new SettingsState()} />
    );

    expect(renderedComponent.state().isOpen).toBe(false);
    renderedComponent.find("NavbarToggler").simulate("click");
    expect(renderedComponent.state().isOpen).toBe(true);
    renderedComponent.find("NavbarToggler").simulate("click");
    expect(renderedComponent.state().isOpen).toBe(false);
  });
});
