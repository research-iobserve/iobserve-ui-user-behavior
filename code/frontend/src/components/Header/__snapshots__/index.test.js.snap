exports[`<Header /> should render 1`] = `
<nav
  className="navbar navbar-toggleable navbar-light bg-faded"
  role="navigation">
  <button
    className="navbar-toggler navbar-toggler-right"
    onClick={[Function]}
    type="button">
    <span
      className="navbar-toggler-icon" />
  </button>
  <a
    className="navbar-brand"
    onClick={[Function]}
    style={Object {}}>
    iObserve UBM Visualization
  </a>
  <div
    className="collapse navbar-collapse"
    style={Object {}}>
    <ul
      className="ml-auto navbar-nav">
      <li
        className="nav-item">
        <a
          className="nav-link"
          onClick={[Function]}
          style={Object {}}>
          Visualization
        </a>
      </li>
      <li
        className="nav-item dropdown">
        <a
          aria-expanded={false}
          aria-haspopup="true"
          className="dropdown-toggle nav-link"
          color="secondary"
          data-toggle="dropdown"
          href="#"
          onClick={[Function]}>
          Editor
        </a>
        <div
          aria-hidden={true}
          className="dropdown-menu dropdown-menu-right"
          role="menu"
          tabIndex="-1">
          <button
            className="dropdown-item"
            onClick={[Function]}
            tabIndex="0">
            <a
              onClick={[Function]}
              style={Object {}}>
              New Application
            </a>
          </button>
          <button
            className="dropdown-item"
            onClick={[Function]}
            tabIndex="0">
            <a
              onClick={[Function]}
              style={Object {}}>
              All Applications
            </a>
          </button>
        </div>
      </li>
      <li
        className="nav-item dropdown">
        <a
          aria-expanded={false}
          aria-haspopup="true"
          className="dropdown-toggle nav-link"
          color="secondary"
          data-toggle="dropdown"
          href="#"
          onClick={[Function]}>
          Settings
        </a>
        <div
          aria-hidden={true}
          className="dropdown-menu dropdown-menu-right"
          role="menu"
          tabIndex="-1">
          <h6
            className="dropdown-header"
            onClick={[Function]}
            tabIndex="-1">
            Layout
          </h6>
          <button
            className="dropdown-item"
            onClick={[Function]}
            tabIndex="0">
            cose
          </button>
          <button
            className="dropdown-item"
            onClick={[Function]}
            tabIndex="0">
            grid
          </button>
          <button
            className="dropdown-item"
            onClick={[Function]}
            tabIndex="0">
            <strong>
              breadthfirst
            </strong>
          </button>
        </div>
      </li>
    </ul>
  </div>
</nav>
`;
